paddle/fluid/imperative/prepared_operator.h

```diff
@@ -133,16 +84,31 @@ NameVarMap<VarType> PrepareData(
           VLOG(3) << "Transform Variable " << var_base->Name() << " from "
                   << kernel_type_for_var << " to " << expected_kernel_key;
           framework::Tensor out;
-          auto tmp_var = std::make_shared<VarType>(var_base->Name());
-          tmp_var->SetType(var_base->Type());
           TransformData(expected_kernel_key, kernel_type_for_var, *tensor,
                         &out);
-          SetTensorToVariable(var_base->Var(), out, tmp_var->MutableVar());
-          var_base = tmp_var;
+          if (NeedTransformDataType(kernel_type_for_var, expected_kernel_key)) {
+            // To avoid NameVarMap copy construction overhead in general
+            // scenarios, if inplace transformed, return original input directly
+            inplace_transform = false;
+            if (tmp_ins.empty()) {
+              tmp_ins = ins;
+            }
+            auto tmp_var = std::make_shared<VarType>(var_base->Name());
+            tmp_var->SetType(var_base->Type());
+            SetTensorToVariable(var_base->Var(), out, tmp_var->MutableVar());
+            tmp_ins[name_pair.first][i] = tmp_var;
+          } else {
+            // if dtype is same, transform inplace will not change the original
+            // value, transform inplace to avoid multiple copy
+            SetTensorToVariable(var_base->Var(), out, var_base->MutableVar());
+          }
         }
       }
     }
   }
+  if (inplace_transform) {
+    return ins;
+  }
   return tmp_ins;
```

Comment:
Maybe you can return NameVarMap<VarType>* to avoid copy.

Reply:
done, thx